### DISCUSSION TITLE
Does Blue Retro work with the Apple Magic Keyboard?

### DISCUSSION THREAD

**Discussion by rs1n** [darthcloud/BlueRetro#573] [Q&A]
I've been using the 1.7.3 Dreamcast-specific firmware and for some reason cannot get my Apple Magic Keyboard to pair with the ESP32. I really want to get back into Typing of the Dead. The ESP32 works fine with a bunch of other controllers (PS3, PS4, XBox One, NES30 Arcade Fight Stick, Switch Joy-cons, Switch Pro controller), but for whatever reason the keyboard just won't pair. Is there something I'm doing wrong (or perhaps not doing) that is preventing this keyboard from pairing? Has anyone else tried this specific keyboard? The model number is A1314 (and yes, it pairs fine with my MacBook Pro and iMac).

**Comment by darthcloud** [comment]
Try what I said in this post:
https://github.com/darthcloud/BlueRetro/discussions/518#discussioncomment-4638022

**Reply by rs1n** [reply]
I tried the steps in the linked comment and got the keyboard working. I thought I'd share my experience:

1. I paired the keyboard on Windows 10. If the keyboard is on, press and hold the power button (opposite of where the batteries are inserted) for about 3-5 seconds. The light should be solid green, then off. Then, turn on the keyboard with the power button but keep it held (after about 3 seconds, the green light should appear, and about 5 seconds later it should blink to denote pairing mode; keep holding the power button down even after the blinking). On Windows, add a new Bluetooth device. Once you see the keyboard appear, release the power button select the keyboard on your computer. It will ask for a pin. I just entered 0000 on the pre-existing keyboard already connected to the computer (not the magic keyboard). Then, when it says "connecting..." type 0000 followed by the ENTER key on the magic keyboard. 
2. On the Dreamcast, using the blueretro.io web interface, I made sure to use the Gamepad/Keyboard preset for Output 2 in the Presets menu. Then under the Advanced menu, I made sure that Output 2 was set to keyboard, and also made sure to select Keyboard for both source (SRC) and destination (DST).
3. After that, I paired the magic keyboard with Blue Retro using the 0000 pin set via Windows.

Typing of the Dead worked "OK" but it seems that my typing is too fast for the communication between keyboard -> Blue Retro -> DC. I had to purposely slow down my typing in order to get every key correctly detected. This can be observed in the Password mode. Using the keyboard to enter random keys, I noticed that if I type two different letters at a 100 words per minute (WPM) rate (e.g. u followed by a), the Dreamcast reads uu instead of ua. If I slow down to about 50 WPM, then it correctly detects as ua. Typing u twice at 100 WPM gives the correct uu output. Oddly, a followed by u, no matter how fast, is properly detected as au. These two were not the only pair of letters that behaved this way (it seemed as if almost any pair were similarly quirky). I don't know if this is the fault of game (don't own an official DC keyboard), my keyboard, or Blue Retro.

Thank you for the help and for such a beautiful project!
